routing/routing_integration_tests/bicycle_route_test.cpp

```diff
@@ -190,41 +267,95 @@ UNIT_TEST(Munich_OnewayBicycle2)
+  CalculateRouteAndTestRouteLength(GetVehicleComponents(VehicleType::Bicycle),
+      mercator::FromLatLon(52.253405,6.182288), {0.0, 0.0},
+      mercator::FromLatLon(52.247599,6.197973), 1440.8 /* expectedRouteMeters */);
+}
+
+UNIT_TEST(Poland_AvoidDetourAndExtraCrossings)
+{
+  // Avoid making a longer cycleway detour which involves extra road crossings.
+  // Same as GraphHopper. Valhalla suggests going by the road (worse for cyclists
+  // really preferring to avoid cars whenever possible, may be preferred
+  // for more aggressive ones depending on light phase to go through one traffic light only).
+  // https://github.com/organicmaps/organicmaps/issues/7954
+  // https://github.com/organicmaps/organicmaps/pull/9692#discussion_r1849627559
+  CalculateRouteAndTestRouteLength(GetVehicleComponents(VehicleType::Bicycle),
+      mercator::FromLatLon(50.08227, 20.03348), {0.0, 0.0},
+      mercator::FromLatLon(50.08253, 20.03191), 157.7 /* expectedRouteMeters */);
+}
+
+UNIT_TEST(Germany_DontAvoidNocyclewayResidential)
+{
+  // No strange detours to avoid a nocycleway residential. Same as GraphHopper.
+  // https://github.com/organicmaps/organicmaps/issues/4059#issuecomment-1399338757
+  CalculateRouteAndTestRouteLength(GetVehicleComponents(VehicleType::Bicycle),
+      mercator::FromLatLon(51.33772, 12.41432), {0.0, 0.0},
+      mercator::FromLatLon(51.33837, 12.40958), 359.495 /* expectedRouteMeters */);
+}
+
+UNIT_TEST(UK_UseNocyclewayTertiary)
+{
+  // A preferred route is through nocycleway tertiary roads. Same as all OSM routers.
+  // The detour is via a shared cycleway and with 130m less alt gain, but is 4km longer.
+  CalculateRouteAndTestRouteLength(GetVehicleComponents(VehicleType::Bicycle),
+      mercator::FromLatLon(51.3826906, -2.3481788), {0.0, 0.0},
+      mercator::FromLatLon(51.3615095, -2.3114138), 4468.83 /* expectedRouteMeters */);
+}
+
+UNIT_TEST(Germany_UseResidential)
+{
+  // No long detour to avoid a short residential.
+  // https://github.com/organicmaps/organicmaps/issues/9330
+  CalculateRouteAndTestRouteLength(GetVehicleComponents(VehicleType::Bicycle),
+      mercator::FromLatLon(48.1464472, 11.5589919), {0.0, 0.0},
+      mercator::FromLatLon(48.1418297, 11.5602123), 614.963 /* expectedRouteMeters */);
 }
 
-// https://github.com/organicmaps/organicmaps/issues/4145
 UNIT_TEST(Belarus_StraightFootway)
 {
+  // Prefer footways over roads in Belarus due to local laws.
+  // https://github.com/organicmaps/organicmaps/issues/4145
   CalculateRouteAndTestRouteLength(GetVehicleComponents(VehicleType::Bicycle),
       mercator::FromLatLon(53.8670285, 30.3162749), {0.0, 0.0},
       mercator::FromLatLon(53.876436, 30.3348084), 1613.34 /* expectedRouteMeters */);
 }
 
 UNIT_TEST(Spain_Madrid_ElevationsDetour)
 {
+  /// @todo(pastk): DELETE: there are no detours here atm, the second route is just a tad longer
+  /// and is of the same alt difference, so this test is currently useless
```

Comment:
AFAIR (a long time ago), this test is based on a real case from the EMail of some bicycle rental app. There was a much bigger detour because of previous bugs in altitude calculations.

Reply:
Ok, but do I understand right you want to keep this test? Why?
I understand the test was useful before, but now its not. If fixes to altitude calculations need to be tested then it should be some altitude calculations unit test, not an integration bicycle routing test, right?

Reply:
So WDYT? Keep it or drop it?

Reply:
I've checked this test. Yes, now it can be replaced with simple:
```
  // Check that OM uses dedicated "Carril bici del Paseo de la Castellana".
  CalculateRouteAndTestRouteLength(GetVehicleComponents(VehicleType::Bicycle),
      mercator::FromLatLon(40.459616, -3.690031), {0.0, 0.0},
      mercator::FromLatLon(40.4403523, -3.69267444), 2283.89 /* expectedRouteMeters */);
```

Before data or elevation fix the r2 route was _completely_ different from r1 route (despite of equal start point and a small offset of the end point). That's why it was introduced.

Reply:
Thanks! Updated.